### PR TITLE
[nextion] Add publish actions

### PR DESCRIPTION
**Pull request by pkejval**
# What does this implement/fix?

Adds YAML actions for publishing states of various Nextion sensors. This functionality was available only in lambdas.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

none

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4368

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [X] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
esphome:
  name: test2
  on_boot: 
    then:
      - sensor.nextion.publish:
          id: nex_float
          state: 1
          publish_state: false
          send_to_nextion: true

      - binary_sensor.nextion.publish:
          id: nex_binary
          state: True

      - text_sensor.nextion.publish:
          id: nex_text
          state: "TEST"

      - switch.nextion.publish:
          id: nex_switch
          state: True

esp32:
  board: wemos_d1_mini32
  framework:
    type: esp-idf

logger:
api:
ota:
  platform: esphome

wifi:
  ssid: !secret wifi
  password: !secret wifi_pass
  reboot_timeout: 0s

external_components:
  - source: github://pr#7646
    components: [ nextion ]

sensor:
  - platform: nextion
    id: nex_float
    nextion_id: nex
    component_name: test.float
    precision: 0

binary_sensor:
  - platform: nextion
    id: nex_binary
    component_name: test.binary

text_sensor:
  - platform: nextion
    id: nex_text
    component_name: test.text

switch:
  - platform: nextion
    id: nex_switch
    component_name: test.switch

uart:
  - id: ser
    rx_pin: GPIO16
    tx_pin: GPIO17
    baud_rate: 115200

display:
  - platform: nextion
    id: nex
    uart_id: ser
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by kbx81 — DISMISSED** [review]
LGTM -- @edwardtfn any thoughts/comments?

**Comment by edwardtfn** [comment]
It looks awesome.

**Comment by kbx81** [comment]
One final thought/request @pkejval -- please update the tests for this component to include these actions.

**Comment by pkejval** [comment]
I wrote some tests but I have **no idea how to run them**. Cannot find any documentation about it. Can you give me advice, please?

**Comment by kbx81** [comment]
From the root of the repo: `./script/test_build_components -e compile -c nextion`

File: tests/components/nextion/test.esp32-ard.yaml
Comment (edwardtfn) [inline tests/components/nextion/test.esp32-ard.yaml:1]: It should be:

```yaml
substitutions:
  tx_pin: '17'
  rx_pin: '16'
```

File: tests/components/nextion/test.esp32-ard.yaml [hunk 1]
```diff
@@ -1,60 +1,6 @@
-wifi:
-  ssid: MySSID
-  password: password1
+substitutions:
+  tx_pin: 17
+  rx_pin: 16
```
Comment (kbx81) [inline tests/components/nextion/test.esp32-ard.yaml]:
```suggestion
  tx_pin: GPIO17
  rx_pin: GPIO16
```

File: tests/components/nextion/test.esp32-c3-ard.yaml [hunk 2]
```diff
@@ -1,60 +1,6 @@
-wifi:
-  ssid: MySSID
-  password: password1
+substitutions:
+  tx_pin: 4
+  rx_pin: 5
```
Comment (kbx81) [inline tests/components/nextion/test.esp32-c3-ard.yaml]:
```suggestion
  tx_pin: GPIO4
  rx_pin: GPIO5
```

File: tests/components/nextion/test.esp32-c3-idf.yaml [hunk 3]
```diff
@@ -1,60 +1,6 @@
-wifi:
-  ssid: MySSID
-  password: password1
+substitutions:
+  tx_pin: 4
+  rx_pin: 5
```
Comment (kbx81) [inline tests/components/nextion/test.esp32-c3-idf.yaml]:
```suggestion
  tx_pin: GPIO4
  rx_pin: GPIO5
```

File: tests/components/nextion/test.esp32-idf.yaml [hunk 4]
```diff
@@ -1,60 +1,6 @@
-wifi:
-  ssid: MySSID
-  password: password1
+substitutions:
+  tx_pin: 17
+  rx_pin: 16
```
Comment (kbx81) [inline tests/components/nextion/test.esp32-idf.yaml]:
```suggestion
  tx_pin: GPIO17
  rx_pin: GPIO16
```

File: tests/components/nextion/test.esp8266-ard.yaml [hunk 5]
```diff
@@ -1,60 +1,6 @@
-wifi:
-  ssid: MySSID
-  password: password1
+substitutions:
+  tx_pin: 4
+  rx_pin: 5
```
Comment (kbx81) [inline tests/components/nextion/test.esp8266-ard.yaml]:
```suggestion
  tx_pin: GPIO4
  rx_pin: GPIO5
```

File: tests/components/nextion/test.rp2040-ard.yaml [hunk 6]
```diff
@@ -1,55 +1,8 @@
-uart:
-  - id: uart_nextion
-    tx_pin: 4
-    rx_pin: 5
-    baud_rate: 115200
+substitutions:
+  tx_pin: 4
+  rx_pin: 5
```
Comment (kbx81) [inline tests/components/nextion/test.rp2040-ard.yaml]:
```suggestion
  tx_pin: GPIO4
  rx_pin: GPIO5
```

**Review by kbx81 — CHANGES_REQUESTED** [review]
Fixes so tests will work

File: tests/components/nextion/common.yaml [hunk 7]
```diff
@@ -0,0 +1,290 @@
+esphome:
+  on_boot:
+    # Binary sensor publish action tests
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        state: True
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        state: True
+        publish_state: True
+        send_to_nextion: True
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        state: True
+        publish_state: False
+        send_to_nextion: True
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        state: True
+        publish_state: True
+        send_to_nextion: False
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        state: True
+        publish_state: False
+        send_to_nextion: False
+
+    # Templated
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        state: !lambda 'return true;'
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+        state: !lambda 'return true;'
+        publish_state: !lambda 'return true;'
+        send_to_nextion: !lambda 'return true;'
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+        state: !lambda 'return true;'
+        publish_state: !lambda 'return false;'
+        send_to_nextion: !lambda 'return true;'
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

File: tests/components/nextion/common.yaml [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+        state: !lambda 'return true;'
+        publish_state: !lambda 'return true;'
+        send_to_nextion: !lambda 'return false;'
+
+    - binary_sensor.nextion.publish:
+        id: r0
```
Comment (kbx81) [inline tests/components/nextion/common.yaml]:
```suggestion
        id: r0_sensor
```

**Review by kbx81 — CHANGES_REQUESTED** [review]
Fix binary sensor IDs in tests

File: tests/components/nextion/test.rp2040-ard.yaml [hunk 17]
```diff
@@ -1,55 +1,8 @@
...
-    on_wake:
-      then:
-        lambda: 'ESP_LOGD("display","Display woke up");'
-    on_setup:
-      then:
-        lambda: 'ESP_LOGD("display","Display setup completed");'
-    on_page:
-      then:
-        lambda: 'ESP_LOGD("display","Display shows new page %u", x);'
+wifi: !remove
```
Comment (kbx81) [inline tests/components/nextion/test.rp2040-ard.yaml]: HTTP client **requires** wifi

Comment (kbx81) [inline tests/components/nextion/test.rp2040-ard.yaml]:
```suggestion
```

**Review by kbx81 — DISMISSED** [review]
LGTM -- thanks!

**Review by kbx81 — APPROVED** [review]

**Pull request merged**